08_cell_line_prediction/nbconverted/lasso_range_gene.py

```diff
@@ -0,0 +1,276 @@
+# 
+# This script plots the detailed results of varying regularization strength (LASSO parameter) for a single gene.
+
+# In[1]:
+
+
+import os
+import itertools as it
+
+import numpy as np
+import pandas as pd
+import matplotlib
+import matplotlib.pyplot as plt
+from matplotlib.lines import Line2D
+from scipy.stats import pearsonr
+import seaborn as sns
+
+import pancancer_evaluation.config as cfg
+import pancancer_evaluation.utilities.data_utilities as du
+import pancancer_evaluation.utilities.analysis_utilities as au
+
+get_ipython().run_line_magic('load_ext', 'autoreload')
+get_ipython().run_line_magic('autoreload', '2')
+
+
+# In[2]:
+
+
+results_dir = os.path.join(
+    cfg.repo_root, '08_cell_line_prediction', 'results', 'tcga_to_ccle'
+)
+
+plot_gene = 'KRAS'
+metric = 'aupr'
+
+
+# ### Get coefficient information for each lasso penalty
+
+# In[3]:
+
+
+nz_coefs_df = []
+
+# get coefficient info for training dataset specified above
+for coef_info in au.generate_nonzero_coefficients_lasso_range(results_dir, gene=plot_gene):
+    (gene,
+     cancer_type,
+     seed,
+     lasso_param,
+     coefs_list) = coef_info
+    for fold_no, coefs in enumerate(coefs_list):
+        nz_coefs_df.append(
+            [gene, cancer_type, lasso_param, seed, fold_no, len(coefs)]
+        )
+        
+nz_coefs_df = pd.DataFrame(
+    nz_coefs_df,
+    columns=['gene', 'cancer_type', 'lasso_param', 'seed', 'fold', 'nz_coefs']
+)
+nz_coefs_df.drop(columns=['cancer_type'], inplace=True)
+nz_coefs_df.lasso_param = nz_coefs_df.lasso_param.astype(float)
+nz_coefs_df = nz_coefs_df[nz_coefs_df.gene == plot_gene].copy()
+nz_coefs_df.head()
+
+
+# In[4]:
+
+
+sns.set({'figure.figsize': (12, 5)})
+
+sns.boxplot(
+    data=nz_coefs_df.sort_values(by=['lasso_param']),
+    x='lasso_param', y='nz_coefs'
+)
+
+# color the boxplot lines/edges rather than the box fill
+# this makes it easier to discern colors at the extremes; i.e. very many or few nonzero coefs
+# https://stackoverflow.com/a/72333641
+ax = plt.gca()
+box_patches = [patch for patch in ax.patches if type(patch) == matplotlib.patches.PathPatch]
+num_patches = len(box_patches)
+lines_per_boxplot = len(ax.lines) // num_patches
+for i, patch in enumerate(box_patches):
+    # set the linecolor on the patch to the facecolor, and set the facecolor to None
+    col = patch.get_facecolor()
+    patch.set_edgecolor(col)
+    patch.set_facecolor('None')
+
+    # each box has associated Line2D objects (to make the whiskers, fliers, etc.)
+    # loop over them here, and use the same color as above
+    for line in ax.lines[i * lines_per_boxplot: (i + 1) * lines_per_boxplot]:
+        line.set_color(col)
+        line.set_mfc(col)  # facecolor of fliers
+        line.set_mec(col)  # edgecolor of fliers
+
+plt.title(f'LASSO parameter vs. number of nonzero coefficients, {plot_gene}')
```

Comment:
Is it weird that there's a huge jump between 10 and 50 and between 100 and 250, but almost none between 50 and 100? Am I overthinking this?

Reply:
I think it's just that for this particular gene the C = 50 and C = 100 models happen to be about the same size, so the performance plateaus between those parameters. This doesn't seem to happen for other genes, for instance here's RB1 where the slope is fairly constant:

![image](https://user-images.githubusercontent.com/2345877/221650429-ca7188d5-2ebe-462a-883d-32158e2fea87.png)

So I think since it's not consistent between genes I'm just going to chalk it up to model fitting randomness, at least for the time being.